File: src/test/java/de/fraunhofer/aisec/crymlin/LSPTest.java

package de.fraunhofer.aisec.crymlin;

import de.fraunhofer.aisec.analysis.server.AnalysisServer;
import de.fraunhofer.aisec.analysis.structures.ServerConfiguration;
import org.eclipse.lsp4j.*;
import org.eclipse.lsp4j.services.LanguageClient;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.net.URL;
import java.nio.file.Paths;
import java.util.concurrent.CompletableFuture;

import static org.junit.jupiter.api.Assertions.*;

/** Tests for the server component of the Language Server Protocol. */
class LSPTest {

	private static String parentFolder;
	private static AnalysisServer server;

	@BeforeAll
	static void setup() throws Exception {
		ClassLoader classLoader = AnalysisServerBotanTest.class.getClassLoader();

		URL resource = classLoader.getResource("unittests/order.mark");
		assertNotNull(resource);
		File markPoC1 = new File(resource.getFile());
		assertNotNull(markPoC1);
		parentFolder = markPoC1.getParent() + File.separator;

		// Start an analysis server
		server = AnalysisServer.builder()
				.config(
					ServerConfiguration.builder()
							.launchConsole(false)
							.launchLsp(true)
							.markFiles(parentFolder)
							.build())
				.build();
		server.start();
	}

	@Test
	void testForbidden() {

		LanguageClient lc = new LanguageClient() {

			@Override
			public void telemetryEvent(Object o) {
			}

			@Override
			public void publishDiagnostics(PublishDiagnosticsParams publishDiagnosticsParams) {
				for (Diagnostic d : publishDiagnosticsParams.getDiagnostics()) {
					System.out.println(d.getMessage());
				}
				// one finding is disabled via comment
				assertEquals(4, publishDiagnosticsParams.getDiagnostics().size());
			}

			@Override
			public void showMessage(MessageParams messageParams) {
			}

			@Override
			public CompletableFuture<MessageActionItem> showMessageRequest(
					ShowMessageRequestParams showMessageRequestParams) {
				return null;
			}

			@Override
			public void logMessage(MessageParams messageParams) {
			}
		};
		server.getLSP().connect(lc);

		TextDocumentItem tdi = new TextDocumentItem();
		tdi.setLanguageId("cpp");
		tdi.setText("dummy");
		tdi.setVersion(-1);
		tdi.setUri(Paths.get(parentFolder, "forbidden.cpp").toUri().toString());
		DidOpenTextDocumentParams params = new DidOpenTextDocumentParams(tdi);

		server.getLSP().getTextDocumentService().didOpen(params);
	}

	@Test
	void testOrder() {

		LanguageClient lc = new LanguageClient() {

			@Override
			public void telemetryEvent(Object o) {
			}

			@Override
			public void publishDiagnostics(PublishDiagnosticsParams publishDiagnosticsParams) {
				assertEquals(8, publishDiagnosticsParams.getDiagnostics().size());
			}

			@Override
			public void showMessage(MessageParams messageParams) {
			}

			@Override
			public CompletableFuture<MessageActionItem> showMessageRequest(
					ShowMessageRequestParams showMessageRequestParams) {
				return null;
			}

			@Override
			public void logMessage(MessageParams messageParams) {
			}
		};
		server.getLSP().connect(lc);

		TextDocumentItem tdi = new TextDocumentItem();
		tdi.setLanguageId("cpp");
		tdi.setText("dummy");
		tdi.setVersion(-1);
		tdi.setUri(Paths.get(parentFolder, "order.cpp").toUri().toString());
		DidOpenTextDocumentParams params = new DidOpenTextDocumentParams(tdi);

		server.getLSP().getTextDocumentService().didOpen(params);
	}

	@Test
	void testShutdown() {
		CompletableFuture<Object> shutdownFuture = server.getLSP().shutdown();
		assertTrue(shutdownFuture.isDone()); // expect immediate result
	}
}

File: src/test/java/de/fraunhofer/aisec/codyze/crymlin/LSPTest.kt
package de.fraunhofer.aisec.codyze.crymlin

import de.fraunhofer.aisec.codyze.analysis.AnalysisServer
import de.fraunhofer.aisec.codyze.analysis.ServerConfiguration
import java.io.*
import java.nio.file.Paths
import java.util.concurrent.CompletableFuture
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue
import org.eclipse.lsp4j.DidOpenTextDocumentParams
import org.eclipse.lsp4j.MessageActionItem
import org.eclipse.lsp4j.MessageParams
import org.eclipse.lsp4j.PublishDiagnosticsParams
import org.eclipse.lsp4j.ShowMessageRequestParams
import org.eclipse.lsp4j.TextDocumentItem
import org.eclipse.lsp4j.services.LanguageClient
import org.junit.jupiter.api.*

/** Tests for the server component of the Language Server Protocol. */
internal class LSPTest {
    @Test
    fun testForbidden() {
        val lc: LanguageClient =
            object : LanguageClient {
                override fun telemetryEvent(o: Any) {}
                override fun publishDiagnostics(
                    publishDiagnosticsParams: PublishDiagnosticsParams
                ) {
                    for (d in publishDiagnosticsParams.diagnostics) {
                        println(d.message)
                    }
                    // one finding is disabled via comment
                    assertEquals(4, publishDiagnosticsParams.diagnostics.size)
                }

                override fun showMessage(messageParams: MessageParams) {}
                override fun showMessageRequest(
                    showMessageRequestParams: ShowMessageRequestParams
                ): CompletableFuture<MessageActionItem>? {
                    return null
                }

                override fun logMessage(messageParams: MessageParams) {}
            }
        server.lsp.connect(lc)
        val tdi = TextDocumentItem()
        tdi.languageId = "cpp"
        tdi.text = "dummy"
        tdi.version = -1
        tdi.uri = Paths.get(parentFolder, "forbidden.cpp").toUri().toString()
        val params = DidOpenTextDocumentParams(tdi)
        server.lsp.textDocumentService.didOpen(params)
    }

    @Test
    fun testOrder() {
        val lc: LanguageClient =
            object : LanguageClient {
                override fun telemetryEvent(o: Any) {}
                override fun publishDiagnostics(
                    publishDiagnosticsParams: PublishDiagnosticsParams
                ) {
                    assertEquals(7, publishDiagnosticsParams.diagnostics.size)
                }

                override fun showMessage(messageParams: MessageParams) {}
                override fun showMessageRequest(
                    showMessageRequestParams: ShowMessageRequestParams
                ): CompletableFuture<MessageActionItem>? {
                    return null
                }

                override fun logMessage(messageParams: MessageParams) {}
            }
        server.lsp.connect(lc)
        val tdi = TextDocumentItem()
        tdi.languageId = "cpp"
        tdi.text = "dummy"
        tdi.version = -1
        tdi.uri = Paths.get(parentFolder, "order.cpp").toUri().toString()
        val params = DidOpenTextDocumentParams(tdi)
        server.lsp.textDocumentService.didOpen(params)
    }

    @Test
    fun testShutdown() {
        val shutdownFuture = server.lsp.shutdown()
        assertTrue(shutdownFuture.isDone) // expect immediate result
    }

    companion object {
        private var parentFolder: String? = null
        private lateinit var server: AnalysisServer

        @BeforeAll
        @JvmStatic
        fun setup() {
            val classLoader = AnalysisServerBotanTest::class.java.classLoader
            val resource = classLoader.getResource("unittests/order.mark")
            assertNotNull(resource)

            val markPoC1 = File(resource.file)
            assertNotNull(markPoC1)
            parentFolder = markPoC1.parent + File.separator

            // Start an analysis server
            server =
                AnalysisServer.builder()
                    .config(
                        ServerConfiguration.builder()
                            .launchConsole(false)
                            .launchLsp(true)
                            .markFiles(parentFolder)
                            .build()
                    )
                    .build()
            server.start()
        }
    }
}
